### DISCUSSION TITLE
Loading parameters from pretrained gluon model

### DISCUSSION THREAD

**Discussion by MaJieCornell** [apache/incubator-mxnet#14017] [Q&A]
Hi,

Currently I am using Gluon to do some transfer learning project. What I am trying to do is loading some parameters from a pretrained network to my new model and fine tune on that. However, I just want to load part of the parameters and left some of them to get random initialized. 

In MXNet what I can do is loading the pretrained parameters first and then manually delete some of the parameters I don't want to use from the parameter dictionary. But in gluon seems like we only have interface to load parameters directly from model files and I am not able to delete the parameters at intermediate stage. The only document regarding this issue I find is this: 

https://mxnet.incubator.apache.org/versions/master/tutorials/onnx/fine_tuning_gluon.html

And code example is like:
```
pre_trained = gluon.nn.SymbolBlock(outputs=new_sym, inputs=mx.sym.var('data_0'))
net_params = pre_trained._collect_params_with_prefix()
for param in new_arg_params:
    if param in net_params:
        net_params[param]._load_init(new_arg_params[param], ctx=ctx)
for param in new_aux_params:
     if param in net_params:
          net_params[param]._load_init(new_aux_params[param], ctx=ctx)
```

It will rely on some of the internal function like "_load_init" and "_collect_params_with_prefix". My concern is that will these internal function deprecated and not supported some day? And what is most recommended way of handling it? Since we are using Gluon to build some critical project then we don't want to use some function which will probably not get supported in the future and deprecated silently. 

Thanks,
Jie

**Comment by pengzhao-intel** [comment]
@zhanghang1989 @xinyu-intel to help :)

**Comment by vdantu** [comment]
@mxnet-label-bot add [gluon, question]

**Comment by szha** [comment]
@MaJieCornell you can still update the parameter files by calling `param_dict = mx.nd.load('filename.params')`, which returns a dictionary of NDArrays. You can delete some elements from this dictionary, and then `mx.nd.save('new_filename.params', param_dict)`, and load that in Gluon with `allow_missing=True`.

**Comment by MaJieCornell** [comment]
Thanks for the reply. While it will definitely work but seems like load-delete-write-load is over complicated and don't sound like a good way of doing it. It seems like the way MXNet doing this, i.e. firstly load the parameters in and then set the parameters to the model, is better. Because it will give user flexibility to do whatever change they want to the loaded parameters and then set it. Is there other concern that Gluon is removing this kind of intermediate stage of parameter loading? @szha @rishita

**Comment by szha** [comment]
An easier way is to load all parameters and force randomly reinitialize the layers you want to train. For example:
```python
model.load_parameters(...)
model.classifier.initialize('uniform', force_reinit=True)
```

**Comment by Marcovaldong** [comment]
> @MaJieCornell you can still update the parameter files by calling `param_dict = mx.nd.load('filename.params')`, which returns a dictionary of NDArrays. You can delete some elements from this dictionary, and then `mx.nd.save(param_dict, 'new_filename.params')`, and load that in Gluon with `allow_missing=True`.

@szha  It should be 
```
param_dict = mx.nd.load('filename.params')
mx.nd.save('new_filename.params', param_dict)
```

**Comment by szha** [comment]
@Marcovaldong thanks, fixed.
